### PR TITLE
Field-level Redaction

### PR DESCRIPTION
**Pull request by ermontross**
###  Summary
Add the ability to redact field values based on a field name for a specified time range via the manager API. This implemented is heavily based on [OpenSearch's field masking](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/configuration/DlsFlsFilterLeafReader.java). 

The admin specifies a time range and a field name to redact. It redacts on read at the Lucene level by overwriting the reader. Currently, replaces the value of the field specified with `REDACTED`. 

**Left to-dos for the future**
- hash the value instead of replacing with `REDACTED`
- auto-delete redactions whose time range is outside of the data's retention policy
- possible: redact queries as well as reads

### Requirements

* [x] I've read and understood the [Contributing Guidelines](CONTRIBUTING.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

### CODE REVIEW

File: astra/src/main/java/com/slack/astra/metadata/fieldredaction/FieldRedactionMetadataStore.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package com.slack.astra.metadata.fieldredaction;
+
+import com.slack.astra.metadata.core.AstraMetadataStore;
+import org.apache.curator.x.async.AsyncCuratorFramework;
+import org.apache.zookeeper.CreateMode;
+
+public class FieldRedactionMetadataStore extends AstraMetadataStore<FieldRedactionMetadata> {
+  public static final String REDACTED_FIELD_METADATA_STORE_ZK_PATH = "/redaction";
+
+  public FieldRedactionMetadataStore(AsyncCuratorFramework curatorFramework, boolean shouldCache) {
```
[the remark this answers is not in the document]
Comment (ermontross) [reply astra/src/main/java/com/slack/astra/metadata/fieldredaction/FieldRedactionMetadataStore.java:10]: I'm actually not sure. It should be `true` in dev/prod but I wasn't sure if it should always be `true` for tests. Maybe @bryanlb can weigh in here.

**Review by bryanlb — DISMISSED** [review]

File: astra/src/test/java/com/slack/astra/util/TestingFieldRedactionMetadataStore.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package com.slack.astra.util;
+
+import com.slack.astra.metadata.fieldredaction.FieldRedactionMetadata;
+import com.slack.astra.metadata.fieldredaction.FieldRedactionMetadataStore;
+import java.util.List;
+import org.apache.curator.x.async.AsyncCuratorFramework;
+
+public class TestingFieldRedactionMetadataStore extends FieldRedactionMetadataStore {
```
[the remark this answers is not in the document]
Comment (ermontross) [reply astra/src/test/java/com/slack/astra/util/TestingFieldRedactionMetadataStore.java:9]: It is used in [TemporaryLogStoreAndSearcherExtension](https://github.com/slackhq/astra/pull/1122/files#diff-389a7fb6d8d1f5fca02d36e20303f5011059c064e70ed42b40bcbaadd0f89661R128) which is used in many tests. I can inline it there.

File: astra/src/main/java/com/slack/astra/metadata/fieldredaction/FieldRedactionMetadataStore.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ermontross) [reply astra/src/main/java/com/slack/astra/metadata/fieldredaction/FieldRedactionMetadataStore.java:10]: In this case, does the manager need to maintain a cache?

File: astra/src/main/java/com/slack/astra/chunk/ReadOnlyChunkImpl.java [hunk 3]
```diff
@@ -562,6 +572,7 @@ public void close() throws IOException {
       replicaMetadataStore.close();
       snapshotMetadataStore.close();
       searchMetadataStore.close();
+      fieldRedactionMetadataStore.close();
```
[the remark this answers is not in the document]
Comment (ermontross) [reply astra/src/main/java/com/slack/astra/chunk/ReadOnlyChunkImpl.java:575]: Both the `CachingChunkManager` and the `IndexingChunkManager` close the store, so I can delete this.

File: astra/src/main/java/com/slack/astra/metadata/fieldredaction/FieldRedactionMetadataStore.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (bryanlb) [reply astra/src/main/java/com/slack/astra/metadata/fieldredaction/FieldRedactionMetadataStore.java:10]: The manager shouldn't necessarily need to maintain a cache but the predecessors would.

**Pull request merged**